pilot/pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	electionId = "istio-leader"
+)
+
+type LeaderElection struct {
+	namespace string
+	name      string
+	runFns    []func(stop <-chan struct{})
+	client    kubernetes.Interface
+}
+
+// Run will start leader election, calling all runFns when we become the leader.
+func (l *LeaderElection) Run(stop <-chan struct{}, onComplete func()) error {
+	le, err := l.create()
+	if err != nil {
+		return fmt.Errorf("failed to create leader election: %v", err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		le.Run(ctx)
+		// Register that we have finished leader election
+		// In practice, this typically means we are exiting the process, and should drop our lock
+		if onComplete != nil {
+			onComplete()
+		}
+	}()
+	go func() {
+		<-stop
+		cancel()
+	}()
+	return nil
+}
+
+func (l *LeaderElection) create() (*leaderelection.LeaderElector, error) {
+	callbacks := leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(ctx context.Context) {
+			for _, f := range l.runFns {
+				go f(ctx.Done())
+			}
+		},
+		OnStoppedLeading: func() {
```

Comment:
Is there any chance of losing the lock without shutting down? Is this what line 100 below is referring to?

Reply:
There is - at the very least if someone took away our rbac permission for config map, api server went down, etc we would. In this case, the context will be canceled, so the functions will stop (the library does this, not us).

I was confused looking at this original since I thought we just did nothing when we lose the lock, but its already handled. Probably warrants a comment.

The L100 comment is just for if WE explicitly cancel the context, we also drop the lock immediately. This allows another instance to take over immediately when we exit